MULTIARCH-5587: Enabling power arch builds

KaushikOP opened the pull request:
- Enabling Power arch for frontend & backend build for Hello World.

I'm verifying Skupper on Power for [MULTIARCH-5587](https://issues.redhat.com/browse/MULTIARCH-5587)

CC: @prb112, @mjturek@us.ibm.com & @hash-d

KaushikOP committed:
* Enabling power arch builds
* Converting image image builds to use plano tool
* using python v3.14 as alpine version has updated python

Signed-off-by: Kaushik Talathi <kaushik.talathi1@ibm.com>

prb112 commented:
Hey @hash-d or @ajssmith 
Hope all is well. Could you look at this one?
Many thanks, Paul

KaushikOP commented:
Hey @hash-d,

I wanted to confirm if there’s anything else required from my end to move this PR forward.

hash-d reviewed (changes requested):
@KaushikOP, it was missed when the github workflows were added, but the images were originally built by Plano with Podman.  You can see the configuration for them on the files .plano.py under the image subdirectories.

Z did add their arch on there, but the workflow is using its own Docker command instead of calling Plano.  Ideally, the workflow should call Plano, so local and workflow builds use the same build instructions.

In any case, P needs to be added to .plano.py, at the very least.

prb112 commented on backend/.plano.py:
```diff
@@ -19,7 +19,7 @@
 from plano import *
 
-image_tag = "quay.io/skupper/hello-world-backend"
+image_tag = "quay.io/ktalathi/skupper/hello-world-backend"
```
Shouldn't this be reverted back?

KaushikOP replied:
I'm Validating plano with Actions.

KaushikOP commented:
Converting to draft as validations going on.

KaushikOP commented on frontend/Containerfile:
```diff
@@ -28,7 +28,7 @@ FROM --platform=$TARGETPLATFORM mirror.gcr.io/library/python:alpine AS run
 USER fritz
 
-COPY --from=build /usr/local/lib/python3.13/site-packages /usr/local/lib/python3.13/site-packages
+COPY --from=build /usr/local/lib/python3.14/site-packages /usr/local/lib/python3.14/site-packages
```
The latest image `mirror.gcr.io/library/python:alpine` uses **python v3.14**.

hash-d replied:
lgtm

KaushikOP commented on backend/Containerfile:
```diff
@@ -28,7 +28,7 @@ FROM --platform=$TARGETPLATFORM mirror.gcr.io/library/python:alpine AS run
 USER fritz
 
-COPY --from=build /usr/local/lib/python3.13/site-packages /usr/local/lib/python3.13/site-packages
+COPY --from=build /usr/local/lib/python3.14/site-packages /usr/local/lib/python3.14/site-packages
```
The latest image mirror.gcr.io/library/python:alpine uses python v3.14.

KaushikOP commented:
> @KaushikOP, it was missed when the github workflows were added, but the images were originally built by Plano with Podman. You can see the configuration for them on the files .plano.py under the image subdirectories.
> 
> Z did add their arch on there, but the workflow is using its own Docker command instead of calling Plano. Ideally, the workflow should call Plano, so local and workflow builds use the same build instructions.
> 
> In any case, P needs to be added to .plano.py, at the very least.


* Updated workflow files to use plano for image build and push
* Updated individual `.plano.py` files for power arch using appropriate container file
* Updated python version in Containerfile as base image uses updated version.

* Validated on my forked repo with custom image names - https://github.com/KaushikOP/skupper-example-hello-world/actions

KaushikOP commented on .github/workflows/build-multiarch-frontend.yml:
```diff
@@ -20,8 +20,4 @@ jobs:
-            --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le \
-            -t quay.io/skupper/hello-world-frontend \
-            --push -f Containerfile .
+          ./plano --file frontend/.plano.py update-gesso,build,push
```
Using plano available in parent directory and providing appropriate `.plano.py` file to use for targets.

hash-d replied:
Same as mentioned above for backend (running from repo root dir):

```
$ podman run -p 8081:8080 -ti quay.io/skupper/hello-world-frontend 
python: can't open file '/home/fritz/python/main.py': [Errno 2] No such file or directory
```
I tested running from the `frontend` dir with the following change, and the image ran fine:
```
-    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file frontend/Containerfile --manifest {image_tag} .")
+    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file Containerfile --manifest {image_tag} .")
```

I had to set `PYTHONPATH` to `<repo_root>/python` before running it, though.

KaushikOP replied:
Earlier Changes in my custom image had the similar issue; I changed the context for image builds and now I can see image running.

`
$ podman run -p 8080:8080 -ti --entrypoint sh quay.io/ktalathi/skupper/hello-world-frontend:latest
~ $ ls
python  static
~ $ ls -l python/
total 16
-rw-r--r--    1 fritz    root          8682 Nov 18 12:05 animalid.py
-rw-r--r--    1 fritz    root          3536 Nov 18 12:05 main.py
~ $ ls -l static/
total 12
drwxr-xr-x    2 fritz    root           161 Nov 18 12:05 gesso
-rw-r--r--    1 fritz    root           268 Nov 18 12:05 index.html
-rw-r--r--    1 fritz    root           124 Nov 18 12:05 main.css
-rw-r--r--    1 fritz    root          2434 Nov 18 12:05 main.js
~ $
~ $ exit
$

$ podman run quay.io/ktalathi/skupper/hello-world-frontend:latest
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)

^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
$
`

KaushikOP commented on .github/workflows/build-multiarch-backend.yml:
```diff
@@ -20,8 +20,4 @@ jobs:
-            --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le \
-            -t quay.io/skupper/hello-world-backend \
-            --push -f Containerfile . 
+           ./plano --file backend/.plano.py build,push
```
Using plano available in parent directory and providing appropriate `.plano.py` file to use for targets.

prb112 replied:
makes sense

hash-d replied:
@KaushikOP , this looks ok, but have you successfully run an image generated with that command?  I have the impression that, being run from the repo root directory, it's pulling the wrong `python` directory.  When I ran using an image built with these instructions, this is what I got:

```
$ podman run quay.io/skupper/hello-world-backend -p 8080:8080
python: can't open file '/home/fritz/python/main.py': [Errno 2] No such file or directory
```

Inspecting the image, the contents of the `python` directory correspond to the one at the repo root (`plano`, `skewer`), and not the one from `backend/python` (`main.py`, `thingid.py`):

```
$ podman run -p 8080:8080 -ti --entrypoint sh quay.io/skupper/hello-world-backend                                                                                          
~ $ ls
python
~ $ ls -l python
total 8
lrwxrwxrwx    1 fritz    root            31 Aug 22  2024 plano -> ../external/skewer/python/plano
lrwxrwxrwx    1 fritz    root            32 Aug 22  2024 skewer -> ../external/skewer/python/skewer
```

Or perhaps I'm doing something wrong?  If not, the options would be to:

- Keep the `cd` command, run `plano` from the subdir, or
- Update `backend/.plano.py` to use `backend` instead of `.` as the build context directory

I'd rather keep the `cd`, as to not change the current operation on `plano`

KaushikOP replied:
Earlier Changes in my custom image had the similar issue; I changed the context for image builds and now I can see image running.

`$ podman run -p 8080:8080 -ti --entrypoint sh quay.io/ktalathi/skupper/hello-world-frontend:latest
~ $ ls
python  static
~ $ ls -l python/
total 16
-rw-r--r--    1 fritz    root          8682 Nov 18 12:05 animalid.py
-rw-r--r--    1 fritz    root          3536 Nov 18 12:05 main.py
~ $ ls -l static/
total 12
drwxr-xr-x    2 fritz    root           161 Nov 18 12:05 gesso
-rw-r--r--    1 fritz    root           268 Nov 18 12:05 index.html
-rw-r--r--    1 fritz    root           124 Nov 18 12:05 main.css
-rw-r--r--    1 fritz    root          2434 Nov 18 12:05 main.js
~ $
~ $ exit
$

$ podman run quay.io/ktalathi/skupper/hello-world-backend:latest
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)


^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
$
`

> Keep the cd command, run plano from the subdir, or

This causes issue as plano is not getting initialized properly in subdir. It only works properly in root dir.
That is why going with other option and setting context in specific `.plano.py` files in sub directories.

KaushikOP commented on frontend/.plano.py:
```diff
@@ -30,7 +30,7 @@ def build(no_cache=False):
     run(f"podman manifest rm {image_tag}", check=False)
     run(f"podman rmi {image_tag}", check=False)
-    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x --manifest {image_tag} .")
+    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file frontend/Containerfile --manifest {image_tag} .")
```
Added Power arch and context for Containerfile to use for build

KaushikOP commented on backend/.plano.py:
```diff
@@ -29,7 +29,7 @@ def build(no_cache=False):
     run(f"podman manifest rm {image_tag}", check=False)
     run(f"podman rmi {image_tag}", check=False)
-    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x --manifest {image_tag} .")
+    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file backend/Containerfile --manifest {image_tag} .")
```
Added Power arch and context for Containerfile to use for build

mjturek commented:
@hash-d is there any need to test the artifacts that come out of this before merging it?

hash-d reviewed (changes requested):
@kaushikOp, perhaps I'm doing something wrong? but the images I generated mirroring the modifications on the PR were invalid.  Can  you check my other comments and verify that?

Also, I see that the `plano test` run failed on your fork.  It seems to me that the reason is that the hello-world example is still set for v1 on main, while the `main` workflow (`.github/workflows/main.yaml`) that runs the test is downloading the latest skupper version's CLI, which is v2:

```
  name: main                                                                                                                                                                                                         
  on:                                                                                                                                                                                                                
    (...)                                                                                                                                                                                    
  jobs:                                                                                                                                                                                                              
    test:                                                                                                                                                                                                            
      strategy:                                                                                                                                                                                                      
        (...)
        matrix:                                                                                                                                                                                                      
          skupper-version: [latest, main]                                                                                                                                                                            
      steps:                                                                                                                                                                                                         
        - uses: actions/checkout@v4                                                                                                                                                                                  
        (...)
        - run: curl https://skupper.io/install.sh | bash -s -- --version ${{matrix.skupper-version}}                                                                                                                 
```

I think the only way around this at this point is to change the versions in the matrix, removing `main` and `latest` and setting some specific versions (say, 1.9.4 and 1.8.7, the latest 1.x releases), while the example itself is not updated to v2.

KaushikOP commented on backend/.plano.py:
```diff
@@ -29,7 +29,7 @@ def build(no_cache=False):
     run(f"podman manifest rm {image_tag}", check=False)
     run(f"podman rmi {image_tag}", check=False)
-    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x --manifest {image_tag} .")
+    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file backend/Containerfile --manifest {image_tag} ./backend")
```
> Keep the cd command, run plano from the subdir, or
> Update backend/.plano.py to use backend instead of . as the build context directory

Updated to use appropriate context image build.
The plano tool is directly available only at root dir.  Plus, the dependant python packages in root python dir.

KaushikOP commented on frontend/.plano.py:
```diff
@@ -30,7 +30,7 @@ def build(no_cache=False):
     run(f"podman manifest rm {image_tag}", check=False)
     run(f"podman rmi {image_tag}", check=False)
-    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x --manifest {image_tag} .")
+    run(f"podman build {no_cache_arg} --format docker --platform linux/amd64,linux/arm64,linux/s390x,linux/ppc64le --file frontend/Containerfile --manifest {image_tag} ./frontend")
```
> Keep the cd command, run plano from the subdir, or
> Update backend/.plano.py to use backend instead of . as the build context directory

Updated to use appropriate context image build.
The plano tool is directly available only at root dir. Plus, the dependant python packages in root python dir.

KaushikOP commented:
Updated the build context and I'm able to view the proper contents and able to run the images fine.
Backend - https://github.com/skupperproject/skupper-example-hello-world/pull/22#discussion_r2538098663
Frontend - https://github.com/skupperproject/skupper-example-hello-world/pull/22#discussion_r2538098045

Here's the run details on my fork.
Backend - https://github.com/KaushikOP/skupper-example-hello-world/actions/runs/19465529084
Frontend - https://github.com/KaushikOP/skupper-example-hello-world/actions/runs/19465529091
Main - https://github.com/KaushikOP/skupper-example-hello-world/actions/runs/19465529099

Images created - 
Backend - quay.io/ktalathi/skupper/hello-world-backend:latest
Frontend - quay.io/ktalathi/skupper/hello-world-frontend:latest

hash-d reviewed (approved):
@KaushikOP, I'm approving this PR, but I will not merge it yet, because it will not pass the tests.

As mentioned  previously, the example is still v1-based, but the CI is installing the latest skupper CLI, which currently is v2, and these two versions are incompatible, causing the test to fail.

I have created a pair of PRs [1] to allow `install.sh` to install the latest v1 version, but I do not know when those will be reviewed and whether they will be merged.

Meanwhile, if you want to have this merged ASAP, here are some options:

- Remove `latest` from the matrix, and put a hardcoded version (such as 1.9.4)
- Remove `latest` from the matrix, and put `v1-dev-release`, instead.
- Wait for one of my `install.sh` PRs to be merged, and update the matrix accordingly

Hardcoding is never good, but it should probably be fine for v1 at this point, and it would be fixed whenever the example/test is updated for v2.

Using `v1-dev-release` might be preferable, but we'd be going from released to unreleased cli in the test, which may cause CI failures.

If you're uncomfortable making these changes, let me know and I'll do them, and you can rebase later.

[1] https://github.com/skupperproject/skupper-website/pull/109 and https://github.com/skupperproject/skupper-website/pull/110

KaushikOP committed:
Update skupper-version for v1 in workflow matrix

pwright commented:
This PR will need to move to be merged against v1 branch, as I plan to move main to reflect the v2 experience

KaushikOP commented:
> @KaushikOP, I'm approving this PR, but I will not merge it yet, because it will not pass the tests.
> 
> As mentioned previously, the example is still v1-based, but the CI is installing the latest skupper CLI, which currently is v2, and these two versions are incompatible, causing the test to fail.
> 
> I have created a pair of PRs [1] to allow `install.sh` to install the latest v1 version, but I do not know when those will be reviewed and whether they will be merged.
> 
> Meanwhile, if you want to have this merged ASAP, here are some options:
> 
> * Remove `latest` from the matrix, and put a hardcoded version (such as 1.9.4)
> * Remove `latest` from the matrix, and put `v1-dev-release`, instead.
> * Wait for one of my `install.sh` PRs to be merged, and update the matrix accordingly
> 
> Hardcoding is never good, but it should probably be fine for v1 at this point, and it would be fixed whenever the example/test is updated for v2.
> 
> Using `v1-dev-release` might be preferable, but we'd be going from released to unreleased cli in the test, which may cause CI failures.
> 
> If you're uncomfortable making these changes, let me know and I'll do them, and you can rebase later.
> 
> [1] [skupperproject/skupper-website#109](https://github.com/skupperproject/skupper-website/pull/109) and [skupperproject/skupper-website#110](https://github.com/skupperproject/skupper-website/pull/110)

@hash-d , updated with `v1-dev-release` and verified at my feature branch here - https://github.com/KaushikOP/skupper-example-hello-world/actions/runs/20064204338

mjturek commented:
New PR targetting the new branch structure available here https://github.com/skupperproject/skupper-example-hello-world/pull/23
